src/psfmachine/machine.py

```diff
@@ -837,289 +635,197 @@ def build_time_model(self, plot=False, downsample=False, split_time_model=False)
-        for bk in range(len(splits) - 1):
-            # find the right mask that select the binned times andd flux to fit the
-            # time model
-            seg_mask = (time_binned[:, 0] >= time_original[splits[bk]]) & (
-                time_binned[:, 0] < time_original[splits[bk + 1] - 1]
+            if self.time_corrector == "pos_corr" and hasattr(self, "pos_corr1"):
+                mpc1 = np.nanmedian(self.pos_corr1, axis=0)
+                mpc2 = np.nanmedian(self.pos_corr2, axis=0)
+                # other_vectors =
+            elif self.time_corrector == "centroid" and hasattr(self, "ra_centroid"):
+                mpc1 = self.ra_centroid.to("arcsec").value / 4
+                mpc2 = self.dec_centroid.to("arcsec").value / 4
+
+            # smooth the vectors
+            mpc1_smooth, mpc2_smooth = smooth_vector(
+                [mpc1, mpc2], splits=splits, filter_size=self.poscorr_filter_size
             )
-            # need to rebuild the A3 DM to use the rigth times/poscorrs
-            A2 = sparse.vstack([A_c] * time_binned[seg_mask].shape[0], format="csr")
-            if hasattr(self, "pos_corr1") and self.time_corrector in [
-                "pos_corr",
-                "centroid",
-            ]:
-                A3 = _combine_A(
-                    A2, poscorr=[poscorr1_binned[seg_mask], poscorr2_binned[seg_mask]]
-                )
-            else:
-                A3 = _combine_A(A2, time=time_binned[seg_mask])
-
-            # No saturated pixels, 1e5 is a hardcoded value for Kepler.
-            k = (
-                (flux_binned_raw < 1e5).all(axis=0)[None, :]
-                * np.ones(flux_binned_raw[seg_mask].shape, bool)
-            ).ravel()
-            # No faint pixels, 100 is a hardcoded value for Kepler.
-            k &= (
-                (flux_binned_raw[seg_mask] > 100).all(axis=0)[None, :]
-                * np.ones(flux_binned_raw[seg_mask].shape, bool)
-            ).ravel()
-            # No huge variability
-            k &= (
-                (np.abs(flux_binned[seg_mask] - 1) < 1).all(axis=0)[None, :]
-                * np.ones(flux_binned[seg_mask].shape, bool)
-            ).ravel()
-            # No nans
-            k &= np.isfinite(flux_binned[seg_mask].ravel()) & np.isfinite(
-                flux_err_binned[seg_mask].ravel()
+            # normalize components
+            mpc1_smooth = (mpc1_smooth - mpc1_smooth.mean()) / (
+                mpc1_smooth.max() - mpc1_smooth.mean()
             )
+            mpc2_smooth = (mpc2_smooth - mpc2_smooth.mean()) / (
+                mpc2_smooth.max() - mpc2_smooth.mean()
+            )
+            # combine them as the first order
+            other_vectors = [mpc1_smooth, mpc2_smooth, mpc1_smooth * mpc2_smooth]
+
+        # create a 3D perturbation matrix
+        P3 = PerturbationMatrix3D(
+            time=_whitened_time,
+            dx=dx,
+            dy=dy,
+            segments=split_time_segments,
```

Comment:
let's keep these keywords the same across `machine` and `perturbation` otherwise it will get confusing. I don't mind if we change some `PerturbationMatrix` keywords, but we should definitely try to be consistent.

Reply:
Ok, I agree.